src/model/HeatPumpPlantLoopEIRHeating.cpp

```diff
@@ -128,18 +141,75 @@ namespace model {
       return result;
     }
 
+    boost::optional<Node> HeatPumpPlantLoopEIRHeating_Impl::heatRecoveryInletNode() const {
+      boost::optional<Node> result;
+
+      if (auto mo = tertiaryInletModelObject()) {
+        result = mo->optionalCast<Node>();
+      }
+      return result;
+    }
+
+    boost::optional<Node> HeatPumpPlantLoopEIRHeating_Impl::heatRecoveryOutletNode() const {
+      boost::optional<Node> result;
+
+      if (auto mo = tertiaryOutletModelObject()) {
+        result = mo->optionalCast<Node>();
+      }
+      return result;
+    }
+
     bool HeatPumpPlantLoopEIRHeating_Impl::addToNode(Node& node) {
 
+      boost::optional<PlantLoop> t_plantLoop = node.plantLoop();
+
+      // If trying to add to a node that is on the demand side of a plant loop
+      if (t_plantLoop) {
+        if (t_plantLoop->demandComponent(node.handle())) {
+          // If there is already a source side water Plant Loop
+          if (boost::optional<PlantLoop> ssLoop = this->sourceSideWaterLoop()) {
+            // And it's not the same as the node's loop
+            if (t_plantLoop.get() != ssLoop.get()) {
+              // And if there is no generator loop (tertiary)
+              if (!this->heatRecoveryLoop().is_initialized()) {
+                // Then try to add it to the tertiary one
+                LOG(Warn, "Calling addToTertiaryNode to connect it to the tertiary (=Heat Recovery Loop) loop for " << briefDescription());
+                return this->addToTertiaryNode(node);
+              }
+            }
+          }
+        }
+      }
+
+      // All other cases, call the base class implementation
       // call the base class implementation to connect the component
       bool ok = WaterToWaterComponent_Impl::addToNode(node);
 
-      // If there's a secondary plant loop, switch the condenser type to "WaterCooled"
+      // If there's a secondary plant loop, switch the condenser type to "WaterSource"
       if (this->sourceSideWaterLoop()) {
         this->setCondenserType("WaterSource");
       }
       return ok;
     }
 
+    bool HeatPumpPlantLoopEIRHeating_Impl::addToTertiaryNode(Node& node) {
```

Comment:
The IDD clearly states that Heat Recovery `\note Not available with water source condenser type`

So maybe we should try to enforce this?